src/main/java/org/logstash/plugins/inputs/http/NettyHttpServer.java

```diff
@@ -0,0 +1,74 @@
+package org.logstash.plugins.inputs.http;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.*;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.util.concurrent.DefaultEventExecutorGroup;
+import org.logstash.plugins.inputs.http.util.SslSimpleBuilder;
+
+import java.io.Closeable;
+import java.util.concurrent.TimeUnit;
+
+import static org.logstash.plugins.inputs.http.util.DaemonThreadFactory.daemonThreadFactory;
+
+/**
+ * Created by joaoduarte on 11/10/2017.
+ */
+public class NettyHttpServer implements Runnable, Closeable {
+    private final int numProcessors = Runtime.getRuntime().availableProcessors();
+    private final ServerBootstrap serverBootstrap;
+    private final String host;
+    private final int port;
+    private SslSimpleBuilder sslSimpleBuilder;
+
+    private final EventLoopGroup processorGroup =
+            new NioEventLoopGroup(numProcessors, daemonThreadFactory("http-input-processor"));
+
+    private final DefaultEventExecutorGroup executorGroup = new DefaultEventExecutorGroup(4,
+            daemonThreadFactory("handler-executor"), 15,
+            new RejectedRequestHandler());
+
+    public NettyHttpServer(final String host, final int port, final SslSimpleBuilder sslSimpleBuilder)
+    {
+        this.host = host;
+        this.port = port;
+        this.sslSimpleBuilder = sslSimpleBuilder;
+        serverBootstrap = new ServerBootstrap()
+                .group(processorGroup)
```

Comment:
Typically you use a boss group and worker group as illustrated in [this](http://netty.io/4.1/xref/io/netty/example/http/snoop/HttpSnoopServer.html#53) example.  If I understand correctly from past discussions, we aren't doing that here to and pushing the worker group to a child handler, to allow that to throw an exception, then handle that exception to set the 429 response code ?

.... and then the SO_BACKLOG is there as a fail safe to reject after too many, such that SO_BACKLOG should always be higher then 429 threshold ?

Just want to understand what is going on here...

Reply:
> we aren't doing that here to and pushing the worker group to a child handler, to allow that to throw an exception, then handle that exception to set the 429 response code ?

Correct. I played a bit with two groups, and I could not find a way to control the creation of new connections: the first never blocks as its job is just to accept connections and pass them to the second group.
With the single worker we pass the work to the custom executor at the end of the pipeline. If that executor is busy we return 429. If we're too busy returning 429s and the connection requests keep flooding in we have SO_BACKLOG in place to reject connections.